### PR TITLE
Add group name to WeightModifier. Add ability to use custom factor ag…

### PR DESCRIPTION
**Pull request by nadiia-volyk**
…gregator function for weight modifiers groups

BACKGROUND:
- Sometimes need to find max(not min as usual) factor for edge with several WeightModifiers. Need polyline weight modifier

DESCRIPTION:
- Add group name to WeightModifier
- Add ability to use custom factor aggregator function for weight modifiers groups. If an edge meets the condition of several WeightModifier objects they will be grouped by group name and factor aggregator function will be applied to WeightModifiers factors. By default - the lowest factor of group is chosen. Finally, factors of all groups will be multiplied.
- Add `SetWeightModifiersGroupFactorAggregator` method for setting factor aggregator functions for different groups. `AggregateFactorMin`, `AggregateFactorMax`, `AggregateFactorMultiply` can be used.
- Add method for creating polyline weight modifier

REQUIRED:
- [ ] All changes are up to date in `CHANGELOG.md`.

- - -
This change is [Reviewable](https://reviewable.io/reviews/hypar-io/Elements/963)

### CODE REVIEW

**Review by DmytroMuravskyi — COMMENTED** [review]
Don't forget about CHANGELOG.md.

> Reviewed 1 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/hypar-io/Elements/963#-:-NSpK5seDPpNrBOhVWGD:b-yttdh0)* status: 0 of 1 approvals obtained (waiting on @nadiia-volyk)

___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 14 at r1](https://reviewable.io/reviews/hypar-io/Elements/963#-NSpLqcU7BkVoOmhPCCH:-NSpLqcVEr7rFBz_Q-LO:b-bj13xt) ([raw file](https://github.com/hypar-io/Elements/blob/e041b459e0485f42839b6dfeaacc8b762e9d6a46/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L14)):*
> ```Smalltalk
>             new Dictionary<string, WeightModifier>();
> 
>         private Dictionary<string, Func<double, double, double>> _groupFactorAggregators =
> ```

I'm not sure if we need another dictionary. Maybe modifiers just can share a group object?
___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 125 at r1](https://reviewable.io/reviews/hypar-io/Elements/963#-NSpC0yh01wqHLZMaxbQ:-NSpC0yh01wqHLZMaxbR:bbb9oro) ([raw file](https://github.com/hypar-io/Elements/blob/e041b459e0485f42839b6dfeaacc8b762e9d6a46/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L125)):*
> ```Smalltalk
>                 name,
>                 new Func<Vertex, Vertex, bool>((a, b) =>
>                 {
> ```

This function is way to expensive for something that is called for every edge, not every line is dependent on input parameters a and b. 
Try to move as much as possible outside of lambda and just pass precomputed data as capture list. As an alternative, new type can be inherited from WeightModifier
___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 141 at r1](https://reviewable.io/reviews/hypar-io/Elements/963#-NSpJnAx1w4096lOZv49:-NSpJnAx1w4096lOZv4A:b-fy2g8) ([raw file](https://github.com/hypar-io/Elements/blob/e041b459e0485f42839b6dfeaacc8b762e9d6a46/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L141)):*
> ```Smalltalk
>                     var hintLine = new RoutingHintLine(polyline, factor, influenceDistance, true, is2D);
>                     var isAffectedParallely = hintLine.Affects(a.Point, b.Point);
>                     var isIntersected = polyline.Intersects(line, out _, includeEnds: true);
> ```

Check if Line DistanceTo Line is more efficient in this case, as intersect won't work with influence distance.
___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 163 at r1](https://reviewable.io/reviews/hypar-io/Elements/963#-NSpBUxy1W2efrICTGzY:-NSpBUxy1W2efrICTGzZ:blnuyyg) ([raw file](https://github.com/hypar-io/Elements/blob/e041b459e0485f42839b6dfeaacc8b762e9d6a46/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L163)):*
> ```Smalltalk
>         {
>             var modifiersGroups = _weightModifiers.Values.GroupBy(m => m.Group);
>             var modifierFactors = new List<double>();
> ```

The function is called for every edge. There is no reason to do grouping every time.  This can be function that operates on one group of modifiers instead.



_Code quote:_
```Smalltalk
ModifierFactor
```

**Review by DmytroMuravskyi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hypar-io/Elements/963)* status: 0 of 1 approvals obtained (waiting on @nadiia-volyk)

___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 66 at r1](https://reviewable.io/reviews/hypar-io/Elements/963#-NStrXmx7jLDdAYWdut_:-NStrXmx7jLDdAYWduta:bxr5l0x) ([raw file](https://github.com/hypar-io/Elements/blob/e041b459e0485f42839b6dfeaacc8b762e9d6a46/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L66)):*
> ```Smalltalk
>         /// <param name="groupName">Group name.</param>
>         /// <param name="groupFactorAggregator">Factor aggregator function.</param>
>         public void SetWeightModifiersGroupFactorAggregator(string groupName, Func<double, double, double> groupFactorAggregator)
> ```

I think we can drop Factor from name to make it little bit shorter.

**Review by nadiia-volyk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hypar-io/Elements/963)* status: 0 of 1 approvals obtained (waiting on @DmytroMuravskyi)

___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 66 at r1](https://reviewable.io/reviews/hypar-io/Elements/963#-NStrXmx7jLDdAYWdut_:-NSuo3Uu3potyQgmvXwU:b-896fix) ([raw file](https://github.com/hypar-io/Elements/blob/e041b459e0485f42839b6dfeaacc8b762e9d6a46/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L66)):*
<details><summary><i>Previously, DmytroMuravskyi wrote…</i></summary><blockquote>

I think we can drop Factor from name to make it little bit shorter.
</blockquote></details>

Done.
___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 77 at r1](https://reviewable.io/reviews/hypar-io/Elements/963#gh-1165230302:-NSuo46iBBrEODH5w8rj:b-896fix) ([raw file](https://github.com/hypar-io/Elements/blob/e041b459e0485f42839b6dfeaacc8b762e9d6a46/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L77)):*
<details><summary><i>Previously, DmytroMuravskyi wrote…</i></summary><blockquote>

Rename to GetWeightModifiersGroup to be consistent with function above and return IEnumerable and ToList is not required.
</blockquote></details>

Done.
___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 125 at r1](https://reviewable.io/reviews/hypar-io/Elements/963#-NSpC0yh01wqHLZMaxbQ:-NSuo4mi4O8ricq7vx0t:b-896fix) ([raw file](https://github.com/hypar-io/Elements/blob/e041b459e0485f42839b6dfeaacc8b762e9d6a46/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L125)):*
<details><summary><i>Previously, DmytroMuravskyi wrote…</i></summary><blockquote>

This function is way to expensive for something that is called for every edge, not every line is dependent on input parameters a and b. 
Try to move as much as possible outside of lambda and just pass precomputed data as capture list. As an alternative, new type can be inherited from WeightModifier
</blockquote></details>

Done.
___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 141 at r1](https://reviewable.io/reviews/hypar-io/Elements/963#-NSpJnAx1w4096lOZv49:-NSuo5do44tG8FWB0bbJ:b-896fix) ([raw file](https://github.com/hypar-io/Elements/blob/e041b459e0485f42839b6dfeaacc8b762e9d6a46/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L141)):*
<details><summary><i>Previously, DmytroMuravskyi wrote…</i></summary><blockquote>

Check if Line DistanceTo Line is more efficient in this case, as intersect won't work with influence distance.
</blockquote></details>

Done.
___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 163 at r1](https://reviewable.io/reviews/hypar-io/Elements/963#-NSpBUxy1W2efrICTGzY:-NSuo6VBE-GndOO_02yu:b-896fix) ([raw file](https://github.com/hypar-io/Elements/blob/e041b459e0485f42839b6dfeaacc8b762e9d6a46/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L163)):*
<details><summary><i>Previously, DmytroMuravskyi wrote…</i></summary><blockquote>

The function is called for every edge. There is no reason to do grouping every time.  This can be function that operates on one group of modifiers instead.
</blockquote></details>

Done.

File: Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs [hunk 1]
```diff
@@ -51,14 +55,43 @@ public void ClearWeightModifiers()
             _weightModifiers.Clear();
         }
 
+        /// <summary>
+        /// Set factor aggregator function for weight modifiers group.
```
Comment (wynged) [inline Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs:59]: let's describe a little bit more about what these do.  "The aggregator function will be applied to any list of weight modifications that share a group name.  This ensures that if a type of modifier is applied to a single edge more than once we have options for how to apply duplicate modifiers."

File: Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs [hunk 2]
```diff
@@ -67,29 +100,159 @@ public WeightModifier AddPlanarWeightModifier(string name, Plane plane, double f
...
+        public WeightModifier AddPolylineWeightModifier(string name, Polyline polyline, double factor, double influenceDistance, bool is2D, string group = null)
+        {
+            RoutingHintLine hintLine;
+            try
+            {
+                hintLine = new RoutingHintLine(polyline, factor, influenceDistance, true, is2D);
+            }
+            catch
+            {
+                hintLine = null;
```
Comment (wynged) [inline Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs:129]: what can result in us not being able to make a hint line?  should we notify the user somehow that this happened?

File: Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs [hunk 3]
```diff
@@ -67,29 +100,159 @@ public WeightModifier AddPlanarWeightModifier(string name, Plane plane, double f
...
             AddWeightModifier(modifier);
             return modifier;
         }
 
         /// <summary>
-        /// Check if edge passes any modifier check and returns the lowest value among them.
+        /// Group weight modifiers.
+        /// </summary>
+        /// <returns>Weight modifiers grouped by group name.</returns>
+        private IEnumerable<IGrouping<string, WeightModifier>> GroupWeightModifiers()
```
Comment (wynged) [inline Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs:172]: rename to `GroupedWeightModifiers`

**Review by wynged — CHANGES_REQUESTED** [review]

**Review by DmytroMuravskyi — COMMENTED** [review]
> Reviewed 3 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/hypar-io/Elements/963)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @nadiia-volyk)

___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 125 at r1](https://reviewable.io/reviews/hypar-io/Elements/963#-NSpC0yh01wqHLZMaxbQ:-NSzHrnF0c7Qu6bE3H1C:bxdas5h) ([raw file](https://github.com/hypar-io/Elements/blob/e041b459e0485f42839b6dfeaacc8b762e9d6a46/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L125)):*
<details><summary><i>Previously, nadiia-volyk wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Few details:
- Do not null check we don't really expect that it's better to throw so the problem is visible. Also, just return null above if hint is null
- Run Affects first since it doesn't require any additional data. 
___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 217 at r2](https://reviewable.io/reviews/hypar-io/Elements/963#-NSzDYHi9y3JjIhLfL66:-NSzDYHi9y3JjIhLfL67:bojdxqd) ([raw file](https://github.com/hypar-io/Elements/blob/3f0abaad3a396211f9349cb905ad6a50a025e8c0/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L217)):*
> ```Smalltalk
>             }
> 
>             if (!_groupFactorAggregators.TryGetValue(modifiersGroup.Key, out var aggregateFactorFunc))
> ```

Looks good, let's also make this function have two parameters instead of group: IEnumerable<WeightModifier> and AggregateFunction.
And retrieve AggregateFunction in CalculateEdgeInfos.

**Review by DmytroMuravskyi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hypar-io/Elements/963)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @nadiia-volyk)

___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 125 at r1](https://reviewable.io/reviews/hypar-io/Elements/963#-NSpC0yh01wqHLZMaxbQ:-NSzMnmn1udgHB7BiTq0:bgrj54j) ([raw file](https://github.com/hypar-io/Elements/blob/e041b459e0485f42839b6dfeaacc8b762e9d6a46/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L125)):*
<details><summary><i>Previously, DmytroMuravskyi wrote…</i></summary><blockquote>

Few details:
- Do not null check we don't really expect that it's better to throw so the problem is visible. Also, just return null above if hint is null
- Run Affects first since it doesn't require any additional data. 
</blockquote></details>

Also, consider grouping code into something like hintLine.Intersects(a ,b)

**Review by nadiia-volyk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/hypar-io/Elements/963)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @DmytroMuravskyi and @wynged)

___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 125 at r1](https://reviewable.io/reviews/hypar-io/Elements/963#-NSpC0yh01wqHLZMaxbQ:-NTDgL-aCOudWRlVJObe:b-896fix) ([raw file](https://github.com/hypar-io/Elements/blob/e041b459e0485f42839b6dfeaacc8b762e9d6a46/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L125)):*
<details><summary><i>Previously, DmytroMuravskyi wrote…</i></summary><blockquote>

Also, consider grouping code into something like hintLine.Intersects(a ,b)
</blockquote></details>

Done.
___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 59 at r2](https://reviewable.io/reviews/hypar-io/Elements/963#gh-1165781974:-NTDfkhVEci5lDMDs6uV:b-896fix) ([raw file](https://github.com/hypar-io/Elements/blob/3f0abaad3a396211f9349cb905ad6a50a025e8c0/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L59)):*
<details><summary><i>Previously, wynged (Eric Wassail) wrote…</i></summary><blockquote>

let's describe a little bit more about what these do.  "The aggregator function will be applied to any list of weight modifications that share a group name.  This ensures that if a type of modifier is applied to a single edge more than once we have options for how to apply duplicate modifiers."
</blockquote></details>

Done.
___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 129 at r2](https://reviewable.io/reviews/hypar-io/Elements/963#gh-1165787130:-NTDflmh2eSqLKeXEb8E:b-i36ree) ([raw file](https://github.com/hypar-io/Elements/blob/3f0abaad3a396211f9349cb905ad6a50a025e8c0/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L129)):*
<details><summary><i>Previously, wynged (Eric Wassail) wrote…</i></summary><blockquote>

what can result in us not being able to make a hint line?  should we notify the user somehow that this happened?
</blockquote></details>

if the polyline is vertical and needs to be used in 2D comparison, such hint line cannot be created.
___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 172 at r2](https://reviewable.io/reviews/hypar-io/Elements/963#gh-1165788765:-NTDgVWS4BM_2u7nCvxv:b-896fix) ([raw file](https://github.com/hypar-io/Elements/blob/3f0abaad3a396211f9349cb905ad6a50a025e8c0/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L172)):*
<details><summary><i>Previously, wynged (Eric Wassail) wrote…</i></summary><blockquote>

rename to `GroupedWeightModifiers`
</blockquote></details>

Done.
___
*[`Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs` line 217 at r2](https://reviewable.io/reviews/hypar-io/Elements/963#-NSzDYHi9y3JjIhLfL66:-NTDge7xElZTL8KVO4EC:b-896fix) ([raw file](https://github.com/hypar-io/Elements/blob/3f0abaad3a396211f9349cb905ad6a50a025e8c0/Elements/src/Spatial/AdaptiveGrid/AdaptiveGraphRoutingWeightModifier.cs#L217)):*
<details><summary><i>Previously, DmytroMuravskyi wrote…</i></summary><blockquote>

Looks good, let's also make this function have two parameters instead of group: IEnumerable<WeightModifier> and AggregateFunction.
And retrieve AggregateFunction in CalculateEdgeInfos.

</blockquote></details>

Done.

**Review by wynged — APPROVED** [review]
> Reviewed 2 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/hypar-io/Elements/963)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @DmytroMuravskyi and @nadiia-volyk)

**Pull request merged**